tools/flaky-test-reporter/ghutil/ghutil.go

```diff
@@ -0,0 +1,327 @@
+func GetComment(org, repo string, commentID int64) (*github.IssueComment, error) {
+	var res *github.IssueComment
+	_, err := retry(
+		fmt.Sprintf("getting comment '%s %s %d'", org, repo, commentID),
+		maxRetryCount,
+		func() (*github.Response, error) {
+			var resp *github.Response
+			var err error
+			res, resp, err = client.Issues.GetComment(ctx, org, repo, commentID)
+			return resp, err
+		},
+	)
+	return res, err
+}
+
+// CreateComment adds comment to issue
+func CreateComment(org, repo string, number int, commentStr string) error {
+	comment := &github.IssueComment{
+		Body: &commentStr,
+	}
+	_, err := retry(
+		fmt.Sprintf("commenting issue '%s %s %d'", org, repo, number),
+		maxRetryCount,
+		func() (*github.Response, error) {
+			var resp *github.Response
+			var err error
+			_, resp, err = client.Issues.CreateComment(ctx, org, repo, number, comment)
+			return resp, err
+		},
+	)
+	return err
+}
+
+// EditComment edits comment with given comment ID
+func EditComment(org, repo string, commentID int64, commentStr string) error {
+	comment := &github.IssueComment{
+		Body: &commentStr,
+	}
+	_, err := retry(
+		fmt.Sprintf("editing comment '%s %s %d'", org, repo, commentID),
+		maxRetryCount,
+		func() (*github.Response, error) {
+			var resp *github.Response
+			var err error
+			_, resp, err = client.Issues.EditComment(ctx, org, repo, commentID, comment)
+			return resp, err
+		},
+	)
+	return err
+}
+
+// AddLabelsToIssue adds label on issue
+func AddLabelsToIssue(org, repo string, number int, labelStrs []string) error {
+	_, err := retry(
+		fmt.Sprintf("add labels '%v' to '%s %s %d'", labelStrs, org, repo, number),
+		maxRetryCount,
+		func() (*github.Response, error) {
+			var resp *github.Response
+			var err error
+			_, resp, err = client.Issues.AddLabelsToIssue(ctx, org, repo, number, labelStrs)
+			return resp, err
+		},
+	)
+	return err
+}
+
+// RemoveLabelForIssue in case test failed, remove label auto:lessflaky if exists
+func RemoveLabelForIssue(org, repo string, number int, labelStr string) error {
+	_, err := retry(
+		fmt.Sprintf("remove label '%s' from '%s %s %d'", labelStr, org, repo, number),
+		maxRetryCount,
+		func() (*github.Response, error) {
+			var resp *github.Response
+			var err error
+			resp, err = client.Issues.RemoveLabelForIssue(ctx, org, repo, number, labelStr)
+			return resp, err
+		},
+	)
+	return err
+}
+
+func updateIssueState(org, repo string, state IssueStateEnum, number int) error {
+	stateString := string(state)
+	issueRequest := &github.IssueRequest{
+		State: &stateString,
+	}
+	_, err := retry(
+		fmt.Sprintf("applying '%s' action on issue '%s %s %d'", stateString, org, repo, number),
+		maxRetryCount,
+		func() (*github.Response, error) {
+			var resp *github.Response
+			var err error
+			_, resp, err = client.Issues.Edit(ctx, org, repo, number, issueRequest)
+			return resp, err
+		},
+	)
+	return err
+}
+
+func waitForRateReset(r *github.Rate) {
+	if r.Remaining <= tokenReserve {
```

Comment:
I remember that the k8s team had an issue with using too many GitHub tokens, and they implemented some sort of caching that saved them a lot of tokens. I think it would be worth checking what they did and enhance this protection mechanism further, if possible.

Reply:
Asked Cole offline, this caching was implemented here: https://github.com/kubernetes/test-infra/tree/master/ghproxy
Based on what we discussed, this wouldn't help us that much, as it caches GET responses which could be beneficial if there are multiple jobs/plugins querying Github for the same resource at the same time. As this job is a daily job, it's very rare that the GET responses would return the same results on the next day. So the short answer is we don't need this caching right now.